### PR TITLE
LPS-166626 Add the Preview Action to the Version History for Web Content

### PR DESCRIPTION
**Pull request by beltranrengifo**
[Story](https://issues.liferay.com/browse/LPS-161020)
[Subtask](https://issues.liferay.com/browse/LPS-166626)

This depends on code from [this another one](https://github.com/liferay-tango/liferay-portal/pull/2972) ☠️ 
Be aware to rebase both when testing.

## Motivation 
Implement the visualization of the web content in a preview page. Previously the modal option was taken into consideration.
This preview action is shipped along the rest of the actions of each version in a web content asset.

## Solution proposed

It will work the same way as other actions that uses a link, through `navigation` utility from FI.
This PR addresses a thing we miss in previous prs: determine a list of available methods each action can use, this way we can make exceptions.

### Content Dashboard testing

- After rebase [this pr](https://github.com/liferay-tango/liferay-portal/pull/2972), deploy all needed modules
- create a web content and modify it several times, 3, 4...
- navigate to the content dashboard, open the info panel for the item and click versions tab
- in any tab click the kebab menu, then click the preview action
- **expected**: a new tab with the web content preview has been opened
image

### Journal testing

- After rebase [this pr](https://github.com/liferay-tango/liferay-portal/pull/2972), deploy all needed modules
- create a web content and modify it several times, 3, 4...
- navigate to Content & Data > Web Content
- once on the main web content. list, click on the kebab menu on our web content, an actions menu should appear
- then click on the preview action
- **expected**: a new tab with the web content preview has been opened
image

### CODE REVIEW

**Comment by beltranrengifo** [comment]
ci:test:sf

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/VersionActions.tsx [hunk 1]
```diff
@@ -15,11 +15,32 @@
...
+import {navigate} from 'frontend-js-web';
 import React from 'react';
 
-const VersionActions = ({actions}: IProps) => {
-	const handleActionClick = ({url}: {url: string}): void => {
+const availableActionMethods: IAvailableActionMethods = {
+	default: (url: string) => {
+		navigate(url);
+	},
+	expire: (url: string): void => {
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/VersionActions.tsx:25]: We can't couple to code to an action, we should define an standard mechanism to decide the type of action.

Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/VersionActions.tsx:25]: That's a good point @cgoncas 
I was thinking about this yesterday...
Then we need to standardise actions methods between BE/FE, don't you think?

Maybe something like `useNavigate`, `useSubmitForm`, `useWhatever`?

Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/VersionActions.tsx:25]: The type of action is fine so far, what I can not see clearly is how to define a method for each action in the FE

Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/VersionActions.tsx:25]: About:

The type of action is fine so far, what I can not see clearly is how to define a method for each action in the FE

Why can't you use the same approach we use with the ContentDashboardActions?

Thanks!!

Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/VersionActions.tsx:25]: > Hey @beltranrengifo
> 
> I have been creating a proposal for you here: #2972
> 
> The types will be SUBMIT_FORM and NAVIGATE so far
> 
> What di you think?

I think that's very fine 👌

Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/VersionActions.tsx:25]: > About:
> 
> The type of action is fine so far, what I can not see clearly is how to define a method for each action in the FE
> 
> Why can't you use the same approach we use with the ContentDashboardActions?
> 
> Thanks!!

Yeah, with this info we can refactor a bit the proposal and link the action type to the method easily as we did [here](https://github.com/beltranrengifo/liferay-portal/blob/644520aebe80879b37f1fd741d3c1befc072224e/modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/transformers/ActionsComponentPropsTransformer.js#L165) I believe.

Then I wonder if the name of the action in needed anymore 🤔

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by claraizquierdo** [comment]
Hi @beltranrengifo , one question: why are we changing the preview in journal page? That is not in the acceptance criteria and I don't know if it should be.

**Review by claraizquierdo — APPROVED** [review]
Everything works perfect!

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
> > Hi @beltranrengifo , one question: why are we changing the preview in journal page? That is not in the acceptance criteria and I don't know if it should be.
> 
> 
> 
> We probably should update the AC. The new browser tab behaviour seems to be the new default one, so I believe we have to change it also in the journal web module. 
> 
> Cc @cgoncas 

Could we show the title only when the URL us not in POPUP mode?

**Comment by beltranrengifo** [comment]
> > > Hi @beltranrengifo , one question: why are we changing the preview in journal page? That is not in the acceptance criteria and I don't know if it should be.
> > 
> > 
> > We probably should update the AC. The new browser tab behaviour seems to be the new default one, so I believe we have to change it also in the journal web module.
> > Cc @cgoncas
> 
> Could we show the title only when the URL is NOT in POPUP mode?

I think we should be able with little effort @cgoncas 
I'll try and let you know asap

**Comment by beltranrengifo** [comment]
hi @cgoncas !
I see the current `urls` we are using come with `p_p_state=pop_up` so I wonder, do we plan to change this?

**Comment by cgoncas** [comment]
> hi @cgoncas ! I see the current `urls` we are using come with `p_p_state=pop_up` so I wonder, do we plan to change this?

I will change it :)

**Comment by cgoncas** [comment]
> > hi @cgoncas ! I see the current `urls` we are using come with `p_p_state=pop_up` so I wonder, do we plan to change this?
> 
> I will change it :)

I have changed it in https://github.com/liferay-tango/liferay-portal/pull/2999.

**Comment by beltranrengifo** [comment]
> > > hi @cgoncas ! I see the current `urls` we are using come with `p_p_state=pop_up` so I wonder, do we plan to change this?
> > 
> > 
> > I will change it :)
> 
> I have changed it in #2999.

Ok! Let me test it with this code so we can ship this safely

**Comment by beltranrengifo** [comment]
JSP changes sent to Echo [here](https://github.com/liferay-echo/liferay-portal/pull/10230)

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:forward

**Pull request closed without merging**